api/src/main/java/jakarta/data/repository/StaticMetamodel.java

```diff
@@ -0,0 +1,110 @@
+package jakarta.data.repository;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * <p>Annotates a class to serve as a static metamodel for an entity,
+ * enabling type-safe access to the names of entity attributes.</p>
+ *
+ * <p>For example, for the following entity,</p>
+ *
+ * <pre>
+ * &#64;Entity
+ * public class Person {
+ *     &#64;Id
+ *     public long ssn;
+ *
+ *     &#64;Embedded
+ *     public Name name;
+ *
+ *     public int yearOfBirth;
+ * }
+ *
+ * &#64;Embeddable
+ * public class Name {
+ *     public String first;
+ *     public String last;
+ * }
+ * </pre>
+ *
+ * <p>You can define a static metamodel as follows,</p>
+ *
+ * <pre>
+ * &#64;StaticMetamodel(Person.class)
+ * public class Person_ {
+ *     public static volatile ssn; // or id
+ *     public static volatile name;
+ *     public static volatile name_first;
+ *     public static volatile name_last;
+ *     public static volatile yearOfBirth;
+ * }
+ * </pre>
+ *
+ * <p>And use it to refer to entity attributes in a type-safe manner,</p>
+ *
+ * <pre>
+ * pageRequest = Pageable.ofSize(20).sortBy(Sort.asc(Person_.yearOfBirth),
+ *                                          Sort.asc(Person_.name_last),
+ *                                          Sort.asc(Person_.name_first),
+ *                                          Sort.asc(Person_.ssn));
+ * </pre>
+ *
+ * <p>When a class is annotated as a {@code StaticMetamodel}, Jakarta Data providers
+ * that provide a repository for the entity type must assign the value of each field
+ * that meets the following criteria:</p>
+ *
+ * <ul>
+ * <li>The field type is {@link String}.</li>
+ * <li>The field is {@code public}.</li>
+ * <li>The field is {@code static}.</li>
+ * <li>The field is not {@code final}.</li>
+ * <li>The field is {@code volatile}.</li>
+ * <li>The name of the field, ignoring case, matches the name of an entity attribute,
+ * where the {@code _} character delimits the attribute names of hierarchical structures
+ * such as embedded classes.</li>
+ * </ul>
+ *
+ * <p>The Jakarta Data provider assigns the field value to the name of the corresponding
+ * entity attribute.</p>
+ *
+ * <p>Additionally, a field that meets the above criteria except for the name
+ * and is named {@code id} must be assigned by the Jakarta Data provider to the
+ * name of the unique identifier entity attribute if a single entity attribute
+ * represents the unique identifier.</p>
+ *
+ * <p>In cases where multiple Jakarta Data providers provide repositories for the same
+ * entity type, no guarantees are made of the order in which the Jakarta Data providers
+ * assign values to the fields of the class that is annotated with {@code StaticMetamodel}.</p>
+ */
+@Documented
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.TYPE)
+public @interface StaticMetamodel {
```

Comment:
IMHO: it should be compatible not just with the repository domain but with all Data patterns that we might bring in the future; thus, it should be in the data package.

Reply:
That's an excellent comment.  I added a [commit](https://github.com/jakartaee/data/pull/258/commits/7d8b43c226f9f4e3212788a139a0bfc422d9ee83) to move it to the data package.